marathon_acme/clients.py

```diff
@@ -41,37 +79,69 @@ def _log_request_error(self, failure, url):
         log.err(failure, 'Error performing request to %s' % (url,))
         return failure
 
-    def request(self, method, path, query=None, endpoint=None, json_data=None,
-                raise_for_status=False, **kwargs):
+    def _compose_url(self, url, kwargs):
+        """
+        Compose a URL starting with the given URL (or self.url if that URL is
+        None) and using the values in kwargs.
+
+        :param str url:
+            The base URL to use. If None, ``self.url`` will be used instead.
+        :param dict kwargs:
+            A dictionary of values to override in the base URL. Relevant keys
+            will be popped from the dictionary.
+        """
+        if url is None:
+            url = self.url
+
+        compose_kwargs = {}
+        if url is not None:
+            split_result = urisplit(url)
+            compose_kwargs.update({
+                'scheme': split_result.scheme,
+                'host': split_result.host,
+                'port': split_result.port,
+                'path': split_result.path,
+                'query': split_result.query,
+                'fragment': split_result.fragment
+            })
```

Comment:
I haven't been able to find a way without using `getattr`.

It's annoying because `uritools.SplitResult` is a "named tuple with extensions". The tuple contains scheme, authority, path, query and fragment. But we don't want to use the authority -- we want to use its parts (userinfo, host & port) which are only attributes and not in the tuple.